src/mlpack/methods/gmm/diagonal_em_fit_impl.hpp

```diff
@@ -0,0 +1,387 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_GMM_DIAGONAL_EM_FIT_IMPL_HPP
+#define MLPACK_METHODS_GMM_DIAGONAL_EM_FIT_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "diagonal_em_fit.hpp"
+#include "no_constraint.hpp"
+
+namespace mlpack {
+namespace gmm {
+
+//! Constructor.
+template<typename InitialClusteringType, typename CovarianceConstraintPolicy>
+DiagonalEMFit<InitialClusteringType, CovarianceConstraintPolicy>::
+DiagonalEMFit(const size_t maxIterations,
+              const double tolerance,
+              InitialClusteringType clusterer,
+              CovarianceConstraintPolicy constraint) :
+              maxIterations(maxIterations),
+              tolerance(tolerance),
+              clusterer(clusterer),
+              constraint(constraint)
+{ /* Nothing to do. */ }
+
+template<typename InitialClusteringType, typename CovarianceConstraintPolicy>
+void DiagonalEMFit<InitialClusteringType, CovarianceConstraintPolicy>::
+Estimate(const arma::mat& observations,
+         std::vector<distribution::DiagonalGaussianDistribution>& dists,
+         arma::vec& weights,
+         const bool useInitialModel)
+{
+  // If it is not on Windows, we use the Armadillo's gmm_diag class by calling
+  // ArmadilloGMMWrapper().
+  #ifndef _WIN32
+  ArmadilloGMMWrapper(observations, dists, weights, useInitialModel);
+
+  #else
+  if (!useInitialModel)
+    InitialClustering(observations, dists, weights);
+
+  // Set the initial log likelihood.
+  double l = LogLikelihood(observations, dists, weights);
+
+  Log::Debug << "DiagonalGMM::Estimate(): initial clustering log likelihood: "
+      << l << std::endl;
+
+  // Initialize the old log likelihood for comparison later.
+  double lOld = -DBL_MAX;
+
+  // Create the conditional probability matrix.
+  arma::mat condProb(observations.n_cols, dists.size());
+
+  // Iterate to update the model until no more improvement is found.
+  size_t iteration = 1;
+  while (std::abs(l - lOld) > tolerance && iteration != maxIterations)
+  {
+    // E step: Calculate the conditional probabilities of the given
+    // observations choosing a particular gaussian using current parameters.
+    for (size_t k = 0; k < dists.size(); k++)
+    {
+      arma::vec condProbAlias = condProb.unsafe_col(k);
+      dists[k].Probability(observations, condProbAlias);
+      condProbAlias *= weights[k];
+    }
+
+    // Normalize row-wise.
+    for (size_t j = 0; j < condProb.n_rows; j++)
+    {
+      // Avoid dividing by zero.
+      const double probSum = accu(condProb.row(j));
+      if (probSum != 0.0)
+        condProb.row(j) /= probSum;
+    }
+
+    // Store the sum of responsibilities over all the observations.
+    arma::vec N = arma::trans(arma::sum(condProb));
```

Comment:
Can you change the name of this so it doesn't start with a capital letter please? :+1: